_posts/2021-09-21-release-1.14.0.md

```diff
@@ -0,0 +1,274 @@
+where Flink internally does it in milliseconds).
+
+This feature is mainly useful for large platform deployments where Flink runs many batch
+jobs of varying kind and manages the resources actively (dynamically requesting and releasing resources
+from Kubernetes or Yarn, as opposted to opaque reactive deployments). Please refer to the
+[Fine-grained Resource Management Docs](https://nightlies.apache.org/flink/flink-docs-release-1.14/docs/deployment/finegrained_resource/)
+for details on how to use this feature.
+
+<figure style="align-content: center">
+  <img src="{{ site.baseurl }}/img/blog/2021-09-25-release-1.14.0/fine_grained_resource_management.png" style="display: block; margin-left: auto; margin-right: auto; width: 600px"/>
+</figure>
+
+Alibaba's internal SQL platform on top of Flink and has used this mechanism for some years now and
+has seen significantly increase the resource utilization of the cluster. This feature was open sourced
+to help other users that build large platforms on top of Flink to optimize resource footprints.
+
+
+# Connectors
+
+## Connector metrics
+
+Flink 1.14 standardized the metrics for connectors ([FLIP-33](https://cwiki.apache.org/confluence/display/FLINK/FLIP-33%3A+Standardize+Connector+Metrics))
+The community will gradually pull them through all connectors, as we rework them
+onto the new unified APIs over the next releases. In Flink 1.14, we cover the Kafka connector,
+and partially the File connectors.
+
+Connectors are the entry and the exit point for data in Flink job. If a job is not running as
+expected, the connector telemetry is among first parts to check. We believe this will become
+a nice improvement for everyone operating production Flink applications.
+
+## Pulsar Source
+
+Flink added a [new Source and Sink](https://nightlies.apache.org/flink/flink-docs-release-1.14/docs/connectors/datastream/pulsar/).
+for [Apache Pulsar](https://pulsar.apache.org/). Pulsar is a distributed messaging and streaming
+system.
+
+The connector currently supports the DataStream API. Table API/SQL bindings are expected to be
+contributed next.
+
+# Goodbye Legacy SQL Engine
+
+Maintaining an open source project also means saying good-bye to some beloved features eventually.
+When we added the Blink SQL Engine to Flink more than two years ago, it was clear that it would
+eventually replace the previous SQL engine. Blink was faster and more feature complete.
+
+For a year now, Blink has been the default SQL engine. With Flink 1.14 we finally remove all
+code from the previous SQL engine. Not only does this help contributors to be faster with their
+development, but it means we can also drop a plethora of outdated interfaces and reduce confusion
+for users which interfaces to use when implementing custom connectors or functions.
+
+
+# Python
+
+The python API also got more observability features and optimisations.
+
+TBD.
```

Comment:
```suggestion
Python function chaining is a technology which combines multiple Python 
functions into one instance during execution. It could avoid unnecessary 
RPC communications between the Java and Python processes, eliminate 
unnecessary serialization/deserialization overhead and improve the overall 
performance. Python function chaining is already available for Python UDFs 
defined in the Python Table API & SQL. In Flink 1.14, it's also available for the 
Python functions defined in Python DataStream API.

Python functions are executed in a separate Python process in PyFlink. This 
makes it difficult to debug for users although they could use Python's builtin 
logging module in their Python functions. In Flink 1.14, we have addressed 
this problem by introducing the loopback mode which is activated by default in
local deployment. Under loopback mode, user-defined Python functions will be
executed in the Python process of the client side instead of launching new 
Python processes. Users could debug their Python functions easily just by setting
breakpoints in the IDEs before executing the job.

There are also many other improvements, e.g. it has supported to execute PyFlink 
jobs in YARN application mode, supported tgz files as Python archives, allowed to 
configure the Python interpreter of client side, etc. Check out the [relevant docs](https://nightlies.apache.org/flink/flink-docs-release-1.14/docs/dev/python/overview/) 
for details.
```

Reply:
I incorporated (and adjusted) this text.